app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .not_staged
+      .where(username_lower: names.map(&:downcase))
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where(name: names)
+      .index_by(&:name)
+
+    mentionable_group_ids = Group
+      .mentionable(current_user, include_public: false)
+      .where(name: names)
+      .pluck(:id)
+      .to_set
```

Comment:
We should try and avoid having to execute this query every time. Technically, this query should only be executed if `groups` has records. A suggestion I have here to avoid all these conditional is to move most of the logic into a dedicated class and extract each query into a dedicated method which is memoized. That way, the query is only executed when we're sure it is required.